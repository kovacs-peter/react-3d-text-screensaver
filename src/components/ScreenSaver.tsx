import React, { PropsWithChildren, useEffect, useState } from 'react';
import { Environment } from '@react-three/drei';
import { Canvas } from '@react-three/fiber';
import { useIdleTimer } from 'react-idle-timer';
import styled from 'styled-components';
import type { TextProperties } from '../types';
import { Direction } from '../types';
import RotatingText from './RotatingText';

interface ScreenSaverProps {
  timeout?: number; // Seconds after the screensaver activates automatically, can't be used if visible prop is used
  text: string; // Text to rotate
  textProps?: TextProperties;
  visible?: boolean; // If this prop is used the visibility is controlled by the app
  className?: string; // If this is set the default canvas styles will be overwritten
}

const defaultTextProps: TextProperties = {
  direction: Direction.clockwise,
  rotationSpeed: 1.2,
  position: { x: 0, y: 0, z: -30 },
  size: 3,
  height: 1.5,
};

const defaultTimeout = 30;

const CanvasContainer = styled.div`
  width: 100vw;
  height: 100vh;
  z-index: 100000;
  position: fixed;
  top: 0;
  left: 0;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  background-color: #000;
`;

const ScreenSaver = ({
  children,
  className,
  text,
  textProps = defaultTextProps,
  timeout = defaultTimeout,
  visible,
}: PropsWithChildren<ScreenSaverProps>) => {
  const [internalVisible, setInternalVisible] = useState<boolean>(!!visible);

  const { start } = useIdleTimer({
    startOnMount: false,
    onIdle: () => setInternalVisible(true),
    onActive: () => setInternalVisible(false),
    timeout: timeout * 1000,
  });

  useEffect(() => {
    if (typeof visible === 'undefined') start();
  }, []);

  if (internalVisible)
    return (
      <CanvasContainer className={className}>
        <Canvas>
          <Environment background={false} blur={0.8} preset="sunset" />
          <mesh>
            <RotatingText text={text} textProps={textProps} />
          </mesh>
        </Canvas>
        {children}
      </CanvasContainer>
    );
  return null;
};

export default ScreenSaver;
