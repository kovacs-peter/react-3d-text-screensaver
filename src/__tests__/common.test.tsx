import { render } from '@testing-library/react';
import React from 'react';
import 'jest-canvas-mock';
import ScreenSaver from '../components/ScreenSaver';

class ResizeObserver {
  observe() {
    // do nothing
  }
  unobserve() {
    // do nothing
  }
  disconnect() {
    // do nothing
  }
}

window.ResizeObserver = ResizeObserver;

describe('<ScreenSaver />', () => {
  it('renders correctly', () => {
    const { container } = render(<ScreenSaver visible text="test text" />);
    expect(container).toMatchSnapshot();
  });
});
