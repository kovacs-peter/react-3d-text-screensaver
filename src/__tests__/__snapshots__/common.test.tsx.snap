// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ScreenSaver /> renders correctly 1`] = `
<div>
  <div
    class="sc-beySbM jDcxxL"
  >
    <div
      style="position: relative; width: 100%; height: 100%; overflow: hidden; pointer-events: auto;"
    >
      <div
        style="width: 100%; height: 100%;"
      >
        <canvas
          style="display: block;"
        />
      </div>
    </div>
  </div>
</div>
`;
